sopel/plugins/capabilities.py

```diff
@@ -0,0 +1,312 @@
+        """
+        return tuple(sorted(request)) in self._registered
+
+    def is_requested(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is requested.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._requested
+
+    def is_acknowledged(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is acknowledged.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._acknowledged
+
+    def is_denied(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is denied.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._denied
+
+    # register, request, resume, acknowledge, deny, etc.
+
+    def register(self, plugin_name: str, request: capability) -> None:
+        """Register a capability ``request`` for ``plugin_name``.
+
+        :param request: the capability request to register for later
+        """
+        self._registered.setdefault(request.cap_req, {})[plugin_name] = (
+            request, False,
+        )
+        LOGGER.debug('Capability Request registered: %s', str(request))
+
+    def request_available(
+        self,
+        bot: Sopel,
+        available_capabilities: Iterable[str],
+    ) -> None:
+        """Request available capabilities.
+
+        :param bot: the bot instance used to send capability requests
+        :param available_capabilities: available capabilities
+
+        This sends ``CAP REQ`` command for requests that can be made, i.e.
+        all the requested capabilities (with or without prefix) must be
+        available for Sopel to send the request.
+
+        Requests made are stored as requested, other are ignored.
+        """
+        # Requesting capabilities when they are available
+        capabilities = set(available_capabilities)
+        for cap_req, plugin_requests in self._registered.items():
+            # request only if all are available
+            cap_req_enables = {
+                cap.lstrip('-')
+                for cap in cap_req
+            }
+            text = ' '.join(cap_req)
+            if cap_req_enables <= capabilities:
+                self._registered[cap_req] = {
+                    plugin_name: (handler_info[0], False)
+                    for plugin_name, handler_info in plugin_requests.items()
+                }
+                self._requested.add(cap_req)
+                LOGGER.debug('Capability negotiation request: "%s".', text)
+                bot.write(('CAP', 'REQ'), text=text)
```

Comment:
I would like to propose handling `-unadvertised-cap` by just omitting that from the REQ, instead of skipping the request.

A CAP that _can't_ be enabled because the server doesn't support it should yield the same server behavior as explicitly disabling it on a server that _does_ support it.

(Also, `cap_req_enables` is a confusing var name. How about something like `absolute_caps`, after the `abs()` built-in function stripping the sign off a number?)

Reply:
Tested with Libera via telnet (1337 h4x0r skills activated) and indeed, `CAP REQ :account-notify -cap2` (obviously Libera doesn't advertise anything named `cap2`) results in a NAK.

But here's the thing: If a server doesn't even know about a given CAP, it's not going to implement the functionality that CAP would enable, whether or not it's asked to do so. Sending a server that doesn't support `cap2`, `CAP REQ :cap1 cap3`, _should_ yield the same behavior in the end as sending a server that _does_ support it `CAP REQ :cap1 -cap2 cap3`. In both cases (assuming a CAP ACK response), the `cap1` and `cap3` behavior will be active, and the `cap2` behavior will be inactive, just as the `-` in the request demands.

This is the sort of thing that could benefit from asking the IRCv3 folks, though. I'm not up for a conversation on that today, personally, but one of us really should dig into this a bit more before making a decision.

Reply:
I'm really against this, for two reasons:

* This is clearly a edge case, and for CAP, I'm all for accepting limitation on how Sopel interpret what is requested from plugins. If someone wants to remove a Capability while requesting others at the same time, they have the option to split their request, or to accept that their request will be sent "as is", meaning will be accepted or denied based on the specification.
* This is **really* hard to keep track internally. The current tracking is already quite complex because of the protocol, adding yet another level of complexity like "this is my request but you have to map this to other request that contains a specific edge case" is just... hugh... no. I really don't want to do that.

This system is really complex, and I don't want to make it even more complex by having developers getting confused over what is actually send or not to the server: what you see is what you get. The capability is not advertised by the server? The request should not be sent. Otherwise, it can be sent, as it is, and it may or may not fail *all at once*, as specified.

Reply:
Quick note on how to track CAP REQ with ACK/NAK: it is all done with simple string manipulation, there is no way to mark a CAP REQ as "this is request nu°7814" and link that number to a callback handler internally. We have to check if a tuple of strings matches anything in memory: having to keep track of modified CAP REQ is just too much.